src/ProducerEventProcessor/UploaderCaptureEventCollector.cpp

```diff
@@ -0,0 +1,116 @@
+// found in the LICENSE file.
+
+#include "ProducerEventProcessor/UploaderCaptureEventCollector.h"
+
+#include "Introspection/Introspection.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/MakeUniqueForOverwrite.h"
+#include "OrbitBase/ThreadUtils.h"
+#include "capture.pb.h"
+#include "services.grpc.pb.h"
+
+using orbit_capture_uploader::DataReadiness;
+using orbit_grpc_protos::CaptureResponse;
+using orbit_grpc_protos::ClientCaptureEvent;
+
+namespace orbit_producer_event_processor {
+
+UploaderCaptureEventCollector::UploaderCaptureEventCollector() {
+  // Create an output stream which will be used to convert capture events into parts of a well
+  // formatted capture file. This should never fail.
+  output_stream_ =
+      orbit_capture_file::CaptureFileOutputStream::Create(&capture_data_buffer_stream_);
+}
+
+UploaderCaptureEventCollector::~UploaderCaptureEventCollector() {
+  LOG("Total number of events uploaded: %u", total_uploaded_event_count_);
+  LOG("Total number of bytes uploaded: %u", total_uploaded_data_bytes_);
+  LOG("Total number of write event errors: %u", total_write_error_count_);
+
+  if (total_uploaded_event_count_ > 0) {
+    float average_bytes = static_cast<float>(total_uploaded_data_bytes_) /
+                          static_cast<float>(total_uploaded_event_count_);
+    LOG("Average number of bytes per event: %.2f", average_bytes);
+  }
+}
+
+void UploaderCaptureEventCollector::Stop() {
+  absl::MutexLock lock{&mutex_};
+
+  CHECK(output_stream_ != nullptr);
+  auto close_result = output_stream_->Close();
+  if (close_result.has_error()) {
+    LOG("Error while closing output steam: %s", close_result.error().message());
+  }
+}
+
+void UploaderCaptureEventCollector::AddEvent(ClientCaptureEvent&& event) {
+  {
+    absl::MutexLock lock{&mutex_};
+
+    // The output stream will be closed while processing the capture finish event. Drop events
+    // received after closing the output stream.
+    CHECK(output_stream_ != nullptr);
+    if (!output_stream_->IsOpen()) return;
+
+    auto write_result = output_stream_->WriteCaptureEvent(event);
+    if (write_result.has_error()) {
+      total_write_error_count_++;
+      return;
+    }
+    buffered_event_count_++;
+    buffered_event_bytes_ += event.ByteSizeLong();
+  }
+
+  // Close output stream after processing the capture finish event.
+  if (event.event_case() == ClientCaptureEvent::kCaptureFinished) Stop();
+}
+
+DataReadiness UploaderCaptureEventCollector::GetDataReadiness() const {
+  absl::MutexLock lock(&mutex_);
+
+  if (!capture_data_to_upload_.empty()) {
+    return DataReadiness::kHasData;
+
+  } else if (output_stream_->IsOpen() || buffered_event_bytes_ > 0) {
+    return DataReadiness::kWaitingForData;
+
+  } else {
+    return DataReadiness::kEndOfData;
+  }
+}
+
+void UploaderCaptureEventCollector::RefreshUploadDataBuffer() {
+  // Clear capture_data_to_upload_ immediately.
+  capture_data_to_upload_.clear();
```

Comment:
I think that this is related to @beckerhe 's comment about having two methods around this: I guess it's normal that this is not empty because `GetUploadDataBuffer()` only returns a reference to `capture_data_to_upload_`, hence `capture_data_to_upload_` is not cleared.

But it still feels weird. For example, calling this method twice in a row without calling (and using the result of) `GetUploadDataBuffer` causes data loss.

One thing I'm thinking of is making `GetUploadDataBuffer()` return by value `std::move(capture_data_to_upload_)` and clear `capture_data_to_upload_`, but this requires changing `UploadDataInterface` so I'm not sure if it's easily doable.

With such a change I think you might have it as a requirement that `capture_data_to_upload_` be empty here.

Reply:
Thanks for pointing out this!

* About "letting `GetUploadDataBuffer()` return by value `std::move(capture_data_to_upload_)`"

  The `CaptureUploader` uses `libcurl` to transfer data. To be specific, `CaptureUploader` provides curl:
  -  a `ReadCallback` to copy data into curl buffer as well as pause the data transferring session if the `UploadDataInterface::GetDataReadiness` tells `CaptureUploader` to wait for data
  - a `XferInfoCallback` to keep checking data readiness and resume the data transferring session when `UploadDataInterface::GetDataReadiness` tells `CaptureUploader` that there is data to upload.

  The curl buffer has a limited size, this is the reason why we are not making `GetUploadDataBuffer()` return by value `std::move(capture_data_to_upload_)` as this needs the uploader to store data in some other place and then copy data into curl buffer batch-by-batch.

  On possible way is to change the `UploadDataInterface` to have a `TakeData` method which will be called in the `ReadCallback`. And the `TakeData` method will take `curl_buffer_size` bytes data from the upload data buffer.

  This will be considered in a separate PR as we also need to make the change accordingly in google 3.

* About "having it as a requirement that capture_data_to_upload_ be empty here"

  The uploader will call `RefreshUploadDataBuffer` when finishing uploading data in upload data buffer and let the `UploadDataInterface` (i.e., the `UploaderClientCaptureEventCollector` here) to clear and refill the data buffer.
  
  Hence, the `capture_data_to_upload_` will not be empty here as the uploader consumes data just by copying them to the curl buffer. 

 But this will be changed when we change the `UploadDataInterface` to have a `TakeData` method. We don't need to clear the buffer in `TakeData`: each time when the curl call `TakeData` in the uploader's ReadCallback, curl take first several bytes into its own buffer.